File: features/person/src/main/java/com/nivasafinance/features/person/entity/Person.java
package com.nivasafinance.features.person.entity;

import com.nivasafinance.common.audit.AuditableEntity;
import com.nivasafinance.common.dto.AddressData;
import com.nivasafinance.common.dto.IdentifierData;
import com.nivasafinance.features.person.enums.Gender;
import io.hypersistence.utils.hibernate.type.json.JsonType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.Type;
import org.hibernate.type.SqlTypes;
import org.javers.core.metamodel.annotation.TypeName;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;

@Entity
@TypeName("person")
@Table(name = "n_person")
@NoArgsConstructor
@AllArgsConstructor
@Data
@EqualsAndHashCode(callSuper = true)
public class Person extends AuditableEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "first_name", length = 100)
    private String firstName;

    @Column(name = "middle_name", length = 100)
    private String middleName;

    @Column(name = "last_name", length = 100)
    private String lastName;

    @Column(name = "display_name", length = 300)
    private String displayName;

    @Type(JsonType.class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "mobile_numbers", columnDefinition = "jsonb")
    private List<MobileNumberDetails> mobileNumbers;

    @Type(JsonType.class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "address", columnDefinition = "jsonb")
    private List<AddressData> address;

    @Column(name = "date_of_birth")
    private LocalDate dateOfBirth;

    @Column(name = "gender", length = 10)
    @Enumerated(EnumType.STRING)
    private Gender gender;

    @Type(JsonType.class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "identifiers", columnDefinition = "jsonb")
    private List<IdentifierData> identifiers;

    @Type(JsonType.class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "data_ext", columnDefinition = "jsonb")
    private Map<String, Object> extData;

    @Type(JsonType.class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "cb_enquiry_id", columnDefinition = "jsonb")
    private List<Long> cbEnquiryId;

    @Type(JsonType.class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "cb_details", columnDefinition = "jsonb")
    private CreditBureauDetails cbDetails;

    @Type(JsonType.class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "consent_details", columnDefinition = "jsonb")
    private List<ConsentInfo> consentDetails;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ConsentInfo {
        private Long id;
        private String type;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class CreditBureauDetails {
        private Long latestSuccessEnquiryId;
        private Long latestEnquiryId;
    }
}

File: features/person/src/main/java/com/nivasafinance/features/person/entity/Person.kt
package com.nivasafinance.features.person.entity

import com.nivasafinance.common.annotations.NoArg
import com.nivasafinance.common.audit.AuditableEntity
import com.nivasafinance.features.person.enum.Gender
import io.hypersistence.utils.hibernate.type.json.JsonType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import org.hibernate.annotations.JdbcTypeCode
import org.hibernate.annotations.Type
import org.hibernate.type.SqlTypes
import org.javers.core.metamodel.annotation.TypeName
import org.javers.core.metamodel.annotation.Value
import java.time.LocalDate
import java.util.UUID

@Entity
@TypeName("person")
@Table(name = "person")
@NoArg
@Suppress("LongParameterList")
class Person(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    var id: UUID? = null,

    @Column(name = "first_name", length = 100)
    var firstName: String? = null,

    @Column(name = "middle_name", length = 100)
    var middleName: String? = null,

    @Column(name = "last_name", length = 100)
    var lastName: String? = null,

    @Type(JsonType::class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "mobile_numbers", columnDefinition = "jsonb")
    var mobileNumbers: List<MobileNumberDetails>? = null,

    @Column(name = "date_of_birth")
    var dateOfBirth: LocalDate? = null,

    @Column(name = "gender", length = 10)
    var gender: Gender? = null,

    @Type(JsonType::class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "data_ext", columnDefinition = "jsonb")
    var extData: Map<String, Any>? = null

) : AuditableEntity()

@Value
@NoArg
data class MobileNumberDetails(
    var number: String? = null,
    var isPrimary: Boolean? = null,
    var isWhatsappAvailable: Boolean? = null
)
